subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -560,82 +549,51 @@ public void useFailureHandler(TaskFailureHandler handler) {
     }
 
     @Override
-    public boolean executeWithTask(final WorkerLease workerLease, final Action<TaskInternal> taskExecution) {
-        final AtomicReference<TaskInfo> selected = new AtomicReference<TaskInfo>();
-        final AtomicBoolean workRemaining = new AtomicBoolean();
-        coordinationService.withStateLock(new Transformer<ResourceLockState.Disposition, ResourceLockState>() {
-            @Override
-            public ResourceLockState.Disposition transform(ResourceLockState resourceLockState) {
-                if (cancellationToken.isCancellationRequested()) {
-                    if (abortExecution()) {
-                        tasksCancelled = true;
-                    }
-                }
+    @Nullable
+    public TaskInfo selectNextTask(WorkerLeaseRegistry.WorkerLease workerLease) {
+        if (allProjectsLocked()) {
+            return null;
+        }
 
-                workRemaining.set(workRemaining());
-                if (!workRemaining.get()) {
-                    return FINISHED;
-                }
+        Iterator<TaskInfo> iterator = executionQueue.iterator();
+        while (iterator.hasNext()) {
+            TaskInfo taskInfo = iterator.next();
+            if (taskInfo.isReady() && allDependenciesComplete(taskInfo)) {
+                ResourceLock projectLock = getProjectLock(taskInfo);
+                TaskMutationInfo taskMutationInfo = getResolvedTaskMutationInfo(taskInfo);
 
-                if (allProjectsLocked()) {
-                    return RETRY;
+                if (!projectLock.tryLock()) {
+                    continue;
+                }
+                if (!workerLease.tryLock()) {
+                    projectLock.unlock();
+                    continue;
                 }
-
                 try {
-                    selected.set(selectNextTask(workerLease));
+                    // TODO: convert output file checks to a resource lock
+                    if (!canRunWithCurrentlyExecutedTasks(taskInfo, taskMutationInfo)) {
+                        workerLease.unlock();
+                        projectLock.unlock();
+                        continue;
+                    }
                 } catch (Throwable t) {
-                    abortAllAndFail(t);
-                    workRemaining.set(false);
+                    workerLease.unlock();
```

Comment:
Not that this is terrible, but it doesn't feel like this method has gotten simpler by removing the coordination service from this class.  For instance, we now have to track which locks we have acquired as we are looking for tasks and make sure to release them when we discard a task for any reason.  

Another option here would be to pass the `ResourceLockState` in and roll it back rather than track and unlock individual locks.  Then you could pull all of the "conditions" back into a single conditional that either proceeds or rolls back the locks before checking the next task.  Removing the inner call to the coordination service and rolling back the locks using `ResourceLockState` does have the side effect that any other locks acquired inside this call to the coordination service would be rolled back, but we don't currently lock anything outside of this logic, so that's probably not a big deal (i.e. we only used the outer lock to control mutable access to the task execution plan, which _could_ be controlled using another resource lock, but I think it turned out to be slightly more performant to just hold the state lock).  

Another (slightly safer) variation of this would be to surround the call to `selectNextTask()` in a second `withStateLock` call, and pass in _that_ `ResourceLockState` - this would prevent the `ResourceLockState` object from rolling back locks not acquired inside `selectNextTask()`.

Reply:
It's just the two locks here, but I think @wolf's change to use the `ResourceLockState` makes this bit of logic much simpler.